File: ios/scribeai/FirebaseAnalyticsHelper.swift
//
//  FirebaseAnalyticsHelper.swift
//  scribeai
//
//  Firebase Analytics wrapper for tracking DAU/MAU, uninstalls, and key events.
//  Firebase Analytics is free and unlimited.
//
//  Key benefits:
//  - Automatic DAU/MAU tracking
//  - Built-in funnel analysis in Firebase Console
//  - Crashlytics integration
//  - User segmentation
//

import Foundation
import FirebaseCore
import FirebaseAnalytics

class FirebaseAnalyticsHelper {
    static let shared = FirebaseAnalyticsHelper()

    private var isInitialized = false

    private init() {}

    // MARK: - Initialization

    /// Initialize Firebase Analytics. Call once in App init or AppDelegate.
    func initialize() {
        FirebaseApp.configure()
        isInitialized = true
        print("🔥 Firebase Analytics initialized")
    }

    // MARK: - User Properties

    /// Set user ID for cross-device tracking
    func setUserId(_ userId: String?) {
        guard isInitialized else { return }
        Analytics.setUserID(userId)
    }

    /// Set user properties for segmentation
    func setUserProperty(name: String, value: String?) {
        guard isInitialized else { return }
        Analytics.setUserProperty(value, forName: name)
    }

    // MARK: - Standard Events

    /// Log app open - helps with DAU calculation
    func logAppOpen() {
        logEvent(AnalyticsEventAppOpen)
    }

    /// Log screen view
    func logScreenView(screenName: String, screenClass: String? = nil) {
        var params: [String: Any] = [AnalyticsParameterScreenName: screenName]
        if let screenClass = screenClass {
            params[AnalyticsParameterScreenClass] = screenClass
        }
        logEvent(AnalyticsEventScreenView, parameters: params)
    }

    /// Log sign up
    func logSignUp(method: String) {
        logEvent(AnalyticsEventSignUp, parameters: [AnalyticsParameterMethod: method])
    }

    /// Log login
    func logLogin(method: String) {
        logEvent(AnalyticsEventLogin, parameters: [AnalyticsParameterMethod: method])
    }

    // MARK: - Custom Events

    /// Log note created
    func logNoteCreated(sourceType: String) {
        logEvent("note_created", parameters: ["source_type": sourceType])
    }

    /// Log AI content generated
    func logContentGenerated(contentType: String, noteId: String? = nil) {
        var params: [String: Any] = ["content_type": contentType]
        if let noteId = noteId {
            params["note_id"] = noteId
        }
        logEvent("content_generated", parameters: params)
    }

    /// Log subscription events
    func logSubscriptionStarted(productId: String, isTrial: Bool) {
        logEvent("subscription_started", parameters: [
            "product_id": productId,
            "is_trial": isTrial
        ])
    }

    func logSubscriptionCancelled(productId: String, reason: String? = nil) {
        var params: [String: Any] = ["product_id": productId]
        if let reason = reason {
            params["reason"] = reason
        }
        logEvent("subscription_cancelled", parameters: params)
    }

    // MARK: - Review Events

    /// Log review prompt shown
    func logReviewPromptShown() {
        logEvent("review_prompt_shown")
    }

    /// Log review prompt response
    func logReviewPromptResponse(_ response: String) {
        logEvent("review_prompt_response", parameters: ["response": response])
    }

    // MARK: - Retention Events

    /// Log sign out attempt (before retention screen)
    func logSignOutAttempted() {
        logEvent("signout_attempted")
    }

    /// Log sign out cancelled (user stayed)
    func logSignOutCancelled() {
        logEvent("signout_cancelled")
    }

    /// Log sign out completed
    func logSignOutCompleted() {
        logEvent("signout_completed")
    }

    /// Log delete account attempt
    func logDeleteAccountAttempted() {
        logEvent("delete_account_attempted")
    }

    /// Log delete account cancelled (user stayed)
    func logDeleteAccountCancelled() {
        logEvent("delete_account_cancelled")
    }

    /// Log delete account completed with reason
    func logDeleteAccountCompleted(reason: String, notesCount: Int) {
        logEvent("delete_account_completed", parameters: [
            "reason": reason,
            "notes_count": notesCount
        ])
    }

    // MARK: - Core Logging

    private func logEvent(_ name: String, parameters: [String: Any]? = nil) {
        guard isInitialized else {
            print("⚠️ Firebase Analytics not initialized, skipping event: \(name)")
            return
        }

        Analytics.logEvent(name, parameters: parameters)

        #if DEBUG
        print("🔥 Firebase: \(name)")
        #endif
    }
}

File: android/app/src/main/java/com/kreativekoala/scribeai/utils/FirebaseAnalyticsHelper.kt
package com.kreativekoala.scribeai.utils

import android.content.Context
import android.os.Bundle
import android.util.Log
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.ktx.Firebase

/**
 * Firebase Analytics wrapper for tracking DAU/MAU, uninstalls, and key events.
 * Firebase Analytics is free and unlimited.
 *
 * Key benefits:
 * - Automatic DAU/MAU tracking
 * - Uninstall tracking via app_remove event (Android only)
 * - Built-in funnel analysis in Firebase Console
 * - Crashlytics integration
 */
object FirebaseAnalyticsHelper {
    private const val TAG = "FirebaseAnalytics"

    private var firebaseAnalytics: FirebaseAnalytics? = null
    private var isInitialized = false

    /**
     * Initialize Firebase Analytics. Call once in Application or MainActivity.
     */
    fun initialize(context: Context) {
        try {
            firebaseAnalytics = Firebase.analytics
            isInitialized = true
            Log.d(TAG, "Firebase Analytics initialized")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize Firebase Analytics: ${e.message}")
        }
    }

    /**
     * Set user ID for cross-device tracking
     */
    fun setUserId(userId: String?) {
        firebaseAnalytics?.setUserId(userId)
    }

    /**
     * Set user properties for segmentation
     */
    fun setUserProperty(name: String, value: String?) {
        firebaseAnalytics?.setUserProperty(name, value)
    }

    // MARK: - Standard Events

    /**
     * Log app open - helps with DAU calculation
     */
    fun logAppOpen() {
        logEvent(FirebaseAnalytics.Event.APP_OPEN)
    }

    /**
     * Log screen view
     */
    fun logScreenView(screenName: String, screenClass: String? = null) {
        val params = Bundle().apply {
            putString(FirebaseAnalytics.Param.SCREEN_NAME, screenName)
            screenClass?.let { putString(FirebaseAnalytics.Param.SCREEN_CLASS, it) }
        }
        logEvent(FirebaseAnalytics.Event.SCREEN_VIEW, params)
    }

    /**
     * Log sign up
     */
    fun logSignUp(method: String) {
        val params = Bundle().apply {
            putString(FirebaseAnalytics.Param.METHOD, method)
        }
        logEvent(FirebaseAnalytics.Event.SIGN_UP, params)
    }

    /**
     * Log login
     */
    fun logLogin(method: String) {
        val params = Bundle().apply {
            putString(FirebaseAnalytics.Param.METHOD, method)
        }
        logEvent(FirebaseAnalytics.Event.LOGIN, params)
    }

    // MARK: - Custom Events

    /**
     * Log note created
     */
    fun logNoteCreated(sourceType: String) {
        val params = Bundle().apply {
            putString("source_type", sourceType)
        }
        logEvent("note_created", params)
    }

    /**
     * Log AI content generated
     */
    fun logContentGenerated(contentType: String, noteId: String? = null) {
        val params = Bundle().apply {
            putString("content_type", contentType)
            noteId?.let { putString("note_id", it) }
        }
        logEvent("content_generated", params)
    }

    fun logPaywallViewed(source: String? = null) {
        val params = Bundle().apply {
            source?.let { putString("source", it) }
        }
        logEvent("paywall_viewed", params)
    }

    fun logPurchaseCompleted(productId: String, revenue: Double? = null) {
        val params = Bundle().apply {
            putString("product_id", productId)
            revenue?.let { putDouble(FirebaseAnalytics.Param.VALUE, it) }
            putString(FirebaseAnalytics.Param.CURRENCY, "USD")
        }
        logEvent(FirebaseAnalytics.Event.PURCHASE, params)
    }

    /**
     * Log subscription events
     */
    fun logSubscriptionStarted(productId: String, isTrial: Boolean) {
        val params = Bundle().apply {
            putString("product_id", productId)
            putBoolean("is_trial", isTrial)
        }
        logEvent("subscription_started", params)
    }

    fun logSubscriptionCancelled(productId: String, reason: String? = null) {
        val params = Bundle().apply {
            putString("product_id", productId)
            reason?.let { putString("reason", it) }
        }
        logEvent("subscription_cancelled", params)
    }

    // MARK: - Review Events

    /**
     * Log review prompt shown
     */
    fun logReviewPromptShown() {
        logEvent("review_prompt_shown")
    }

    /**
     * Log review prompt response
     */
    fun logReviewPromptResponse(response: String) {
        val params = Bundle().apply {
            putString("response", response) // "positive", "negative", "dismissed"
        }
        logEvent("review_prompt_response", params)
    }

    // MARK: - Retention Events

    /**
     * Log sign out attempt (before retention screen)
     */
    fun logSignOutAttempted() {
        logEvent("signout_attempted")
    }

    /**
     * Log sign out cancelled (user stayed)
     */
    fun logSignOutCancelled() {
        logEvent("signout_cancelled")
    }

    /**
     * Log sign out completed
     */
    fun logSignOutCompleted() {
        logEvent("signout_completed")
    }

    /**
     * Log delete account attempt
     */
    fun logDeleteAccountAttempted() {
        logEvent("delete_account_attempted")
    }

    /**
     * Log delete account cancelled (user stayed)
     */
    fun logDeleteAccountCancelled() {
        logEvent("delete_account_cancelled")
    }

    /**
     * Log delete account completed with reason
     */
    fun logDeleteAccountCompleted(reason: String, notesCount: Int) {
        val params = Bundle().apply {
            putString("reason", reason)
            putInt("notes_count", notesCount)
        }
        logEvent("delete_account_completed", params)
    }

    // MARK: - Core Logging

    private fun logEvent(eventName: String, params: Bundle? = null) {
        if (!isInitialized) {
            Log.w(TAG, "Firebase Analytics not initialized, skipping event: $eventName")
            return
        }

        try {
            firebaseAnalytics?.logEvent(eventName, params)
            Log.d(TAG, "Logged event: $eventName")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to log event $eventName: ${e.message}")
        }
    }
}
